File: app/static/external-models/fbx-global-settings.js
// Ограниченный reader метаданных GlobalSettings бинарного FBX 7.4/7.5.
// НЕ полный FBX-парсер: читает только заголовок и верхнеуровневый узел
// GlobalSettings/Properties70, массивы (Vertices и т.п.) пропускает по
// длине без распаковки, глубина обхода ограничена. Единственная цель —
// независимо от FBXLoader подтвердить единицы и оси перед импортом (см.
// Docs/fbx-ground-implementation-task.md §4). Ошибка формата/версии/
// комбинации осей — явный отказ, не угадывание.
//
// Первая и пока единственная поддерживаемая комбинация осей — профиль
// файла 0507_Aviomotornaya_ZU_5_Ground.fbx: UpAxis=1 (+1), FrontAxis=2
// (+1), CoordAxis=0 (+1). Другие комбинации отклоняются.

const MAX_DEPTH = 6;
const HEADER_MAGIC = "Kaydara FBX Binary  ";

export const SUPPORTED_AXIS_PROFILE = {
  upAxis: 1, upAxisSign: 1,
  frontAxis: 2, frontAxisSign: 1,
  coordAxis: 0, coordAxisSign: 1,
};

class Cursor {
  constructor(view, limit) {
    this.view = view;
    this.offset = 0;
    this.limit = limit;
  }
  need(n) {
    if (this.offset + n > this.limit) throw new RangeError("FBX: выход за границы буфера при чтении");
  }
  u8() { this.need(1); const v = this.view.getUint8(this.offset); this.offset += 1; return v; }
  u32() { this.need(4); const v = this.view.getUint32(this.offset, true); this.offset += 4; return v; }
  u64() { this.need(8); const v = this.view.getBigUint64(this.offset, true); this.offset += 8; return v; }
  i16() { this.need(2); const v = this.view.getInt16(this.offset, true); this.offset += 2; return v; }
  i32() { this.need(4); const v = this.view.getInt32(this.offset, true); this.offset += 4; return v; }
  i64() { this.need(8); const v = this.view.getBigInt64(this.offset, true); this.offset += 8; return v; }
  f32() { this.need(4); const v = this.view.getFloat32(this.offset, true); this.offset += 4; return v; }
  f64() { this.need(8); const v = this.view.getFloat64(this.offset, true); this.offset += 8; return v; }
  bytes(n) { this.need(n); const v = new Uint8Array(this.view.buffer, this.view.byteOffset + this.offset, n); this.offset += n; return v; }
  skip(n) { this.need(n); this.offset += n; }
}

function readProperty(c) {
  const type = String.fromCharCode(c.u8());
  switch (type) {
    case "Y": return c.i16();
    case "C": return c.u8() !== 0;
    case "I": return c.i32();
    case "F": return c.f32();
    case "D": return c.f64();
    case "L": return Number(c.i64());
    case "f": case "d": case "l": case "i": case "b": {
      // Массив: длина/кодировка/сжатая_длина, затем сами данные — ПРОПУСКАЕМ
      // без распаковки: значения GlobalSettings/Properties70 — всегда
      // скаляры, массивы тут не нужны и не читаются намеренно (§4).
      const arrayLen = c.u32();
      const encoding = c.u32();
      const compLen = c.u32();
      if (encoding === 0) {
        const elemSize = { f: 4, d: 8, l: 8, i: 4, b: 1 }[type];
        c.skip(arrayLen * elemSize);
      } else {
        c.skip(compLen);
      }
      return null;
    }
    case "S": case "R": {
      const n = c.u32();
      const raw = c.bytes(n);
      return type === "S" ? new TextDecoder("utf-8", { fatal: false }).decode(raw) : null;
    }
    default:
      throw new Error(`FBX: неизвестный тип свойства '${type}'`);
  }
}

function readNode(c, version, depth) {
  if (depth > MAX_DEPTH) throw new Error("FBX: превышена ограниченная глубина чтения GlobalSettings");
  let endOffset, numProps, propListLen;
  if (version >= 7500) {
    endOffset = Number(c.u64()); numProps = Number(c.u64()); propListLen = Number(c.u64());
  } else {
    endOffset = c.u32(); numProps = c.u32(); propListLen = c.u32();
  }
  const nameLen = c.u8();
  const nameBytes = c.bytes(nameLen);
  const name = new TextDecoder("utf-8", { fatal: false }).decode(nameBytes);
  if (endOffset === 0) return null; // NULL-запись конца списка
  if (endOffset > c.limit) throw new RangeError("FBX: endOffset узла выходит за пределы файла");
  const propsEnd = c.offset + propListLen;
  const props = [];
  for (let i = 0; i < numProps; i++) props.push(readProperty(c));
  if (c.offset > propsEnd) throw new RangeError("FBX: список свойств узла повреждён");
  c.offset = propsEnd;
  const children = [];
  const nullRecSize = version >= 7500 ? 25 : 13;
  while (c.offset < endOffset - nullRecSize) {
    const child = readNode(c, version, depth + 1);
    if (child === null) break;
    children.push(child);
  }
  c.offset = endOffset;
  return { name, props, children };
}

function findChild(node, name) {
  return node.children.find((n) => n.name === name) || null;
}

/**
 * Читает заголовок и GlobalSettings/Properties70 бинарного FBX.
 * @param {ArrayBuffer} buffer
 * @returns {{formatVersion:number, unitScaleFactor:number,
 *   originalUnitScaleFactor:number, upAxis:number, upAxisSign:number,
 *   frontAxis:number, frontAxisSign:number, coordAxis:number,
 *   coordAxisSign:number, mmPerUnit:number}}
 * @throws {Error} явная и понятная причина отказа — не угадывает
 */
export function readFbxGlobalSettings(buffer) {
  const view = new DataView(buffer);
  const headerBytes = new Uint8Array(buffer, 0, Math.min(21, buffer.byteLength));
  const header = new TextDecoder("ascii").decode(headerBytes);
  if (buffer.byteLength < 27 || !header.startsWith("Kaydara FBX Binary")) {
    throw new Error("Файл не является бинарным FBX (нет сигнатуры 'Kaydara FBX Binary'). ASCII FBX не поддерживается.");
  }
  const c = new Cursor(view, buffer.byteLength);
  c.offset = 23;
  const formatVersion = c.u32();
  if (formatVersion < 7400 || formatVersion >= 7700) {
    throw new Error(`Версия формата FBX ${formatVersion} не входит в поддерживаемый диапазон 7400–7699.`);
  }
  c.offset = 27;
  const nullRecSize = formatVersion >= 7500 ? 25 : 13;
  let globalSettingsNode = null;
  while (c.offset < c.limit - nullRecSize) {
    const node = readNode(c, formatVersion, 0);
    if (node === null) break;
    if (node.name === "GlobalSettings") { globalSettingsNode = node; break; }
    if (node.name === "Objects" || node.name === "Connections") break; // GlobalSettings всегда раньше — дальше не ищем
  }
  if (!globalSettingsNode) {
    throw new Error("В файле не найден узел GlobalSettings.");
  }
  const p70 = findChild(globalSettingsNode, "Properties70");
  if (!p70) throw new Error("В GlobalSettings нет Properties70.");
  const settings = {};
  for (const p of p70.children) {
    if (p.name !== "P" || !p.props.length) continue;
    settings[p.props[0]] = p.props.slice(4);
  }
  const num = (key, fallback) => {
    const arr = settings[key];
    return arr && arr[0] != null ? Number(arr[0]) : fallback;
  };
  const result = {
    formatVersion,
    unitScaleFactor: num("UnitScaleFactor", null),
    originalUnitScaleFactor: num("OriginalUnitScaleFactor", null),
    upAxis: num("UpAxis", null),
    upAxisSign: num("UpAxisSign", null),
    frontAxis: num("FrontAxis", null),
    frontAxisSign: num("FrontAxisSign", null),
    coordAxis: num("CoordAxis", null),
    coordAxisSign: num("CoordAxisSign", null),
  };
  if (result.unitScaleFactor == null || !Number.isFinite(result.unitScaleFactor) || result.unitScaleFactor <= 0) {
    throw new Error("GlobalSettings.UnitScaleFactor отсутствует или некорректен.");
  }
  for (const key of ["upAxis", "upAxisSign", "frontAxis", "frontAxisSign", "coordAxis", "coordAxisSign"]) {
    if (result[key] == null) throw new Error(`GlobalSettings.${key} отсутствует.`);
  }
  result.mmPerUnit = result.unitScaleFactor * 10; // FbxSystemUnit: единица файла в см, см→мм = ×10
  return result;
}

/** Отклоняет всё, кроме единственного подтверждённого профиля осей (§2, §4). */
export function assertSupportedAxisProfile(settings) {
  const p = SUPPORTED_AXIS_PROFILE;
  if (
    settings.upAxis !== p.upAxis || settings.upAxisSign !== p.upAxisSign ||
    settings.frontAxis !== p.frontAxis || settings.frontAxisSign !== p.frontAxisSign ||
    settings.coordAxis !== p.coordAxis || settings.coordAxisSign !== p.coordAxisSign
  ) {
    throw new Error(
      "Комбинация осей FBX не поддерживается в этой версии импорта " +
      `(UpAxis=${settings.upAxis}/${settings.upAxisSign}, FrontAxis=${settings.frontAxis}/${settings.frontAxisSign}, ` +
      `CoordAxis=${settings.coordAxis}/${settings.coordAxisSign}). ` +
      "Поддержан только профиль первого приёмочного файла.",
    );
  }
}

File: app/fbx_global_settings.py
"""Серверный аналог app/static/external-models/fbx-global-settings.js —
ограниченный reader GlobalSettings бинарного FBX 7.4/7.5, БЕЗ зависимостей
(вызывается из app/external_models.py). Те же ограничения: глубина, только
верхний узел GlobalSettings/Properties70, массивы пропускаются по длине без
распаковки. Не доверяет тому, что прислал браузер, — независимая проверка
единиц/осей на сервере (см. §7 задания).
"""
import struct
from dataclasses import dataclass
from typing import Optional

MAX_DEPTH = 6

SUPPORTED_AXIS_PROFILE = {
    "up_axis": 1, "up_axis_sign": 1,
    "front_axis": 2, "front_axis_sign": 1,
    "coord_axis": 0, "coord_axis_sign": 1,
}


class FbxGlobalSettingsError(ValueError):
    pass


@dataclass
class Cursor:
    buf: bytes
    offset: int = 0

    def need(self, n):
        if self.offset + n > len(self.buf):
            raise FbxGlobalSettingsError("FBX: выход за границы буфера")

    def u8(self):
        self.need(1); v = self.buf[self.offset]; self.offset += 1; return v

    def u32(self):
        self.need(4); v = struct.unpack_from("<I", self.buf, self.offset)[0]; self.offset += 4; return v

    def u64(self):
        self.need(8); v = struct.unpack_from("<Q", self.buf, self.offset)[0]; self.offset += 8; return v

    def i16(self):
        self.need(2); v = struct.unpack_from("<h", self.buf, self.offset)[0]; self.offset += 2; return v

    def i32(self):
        self.need(4); v = struct.unpack_from("<i", self.buf, self.offset)[0]; self.offset += 4; return v

    def i64(self):
        self.need(8); v = struct.unpack_from("<q", self.buf, self.offset)[0]; self.offset += 8; return v

    def f32(self):
        self.need(4); v = struct.unpack_from("<f", self.buf, self.offset)[0]; self.offset += 4; return v

    def f64(self):
        self.need(8); v = struct.unpack_from("<d", self.buf, self.offset)[0]; self.offset += 8; return v

    def bytes(self, n):
        self.need(n); v = self.buf[self.offset:self.offset + n]; self.offset += n; return v

    def skip(self, n):
        self.need(n); self.offset += n


def _read_property(c: Cursor):
    t = chr(c.u8())
    if t == "Y": return c.i16()
    if t == "C": return c.u8() != 0
    if t == "I": return c.i32()
    if t == "F": return c.f32()
    if t == "D": return c.f64()
    if t == "L": return c.i64()
    if t in "fdlib":
        array_len = c.u32(); encoding = c.u32(); comp_len = c.u32()
        if encoding == 0:
            elem = {"f": 4, "d": 8, "l": 8, "i": 4, "b": 1}[t]
            c.skip(array_len * elem)
        else:
            c.skip(comp_len)
        return None
    if t == "S":
        n = c.u32(); raw = c.bytes(n)
        return raw.decode("utf-8", "replace")
    if t == "R":
        n = c.u32(); c.bytes(n); return None
    raise FbxGlobalSettingsError(f"FBX: неизвестный тип свойства {t!r}")


def _read_node(c: Cursor, version: int, depth: int):
    if depth > MAX_DEPTH:
        raise FbxGlobalSettingsError("FBX: превышена ограниченная глубина")
    if version >= 7500:
        end_offset, num_props, prop_list_len = c.u64(), c.u64(), c.u64()
    else:
        end_offset, num_props, prop_list_len = c.u32(), c.u32(), c.u32()
    name_len = c.u8()
    name = c.bytes(name_len).decode("utf-8", "replace")
    if end_offset == 0:
        return None
    if end_offset > len(c.buf):
        raise FbxGlobalSettingsError("FBX: endOffset вне файла")
    props_end = c.offset + prop_list_len
    props = []
    for _ in range(num_props):
        props.append(_read_property(c))
    if c.offset > props_end:
        raise FbxGlobalSettingsError("FBX: список свойств повреждён")
    c.offset = props_end
    children = []
    null_rec = 25 if version >= 7500 else 13
    while c.offset < end_offset - null_rec:
        child = _read_node(c, version, depth + 1)
        if child is None:
            break
        children.append(child)
    c.offset = end_offset
    return {"name": name, "props": props, "children": children}


def read_fbx_global_settings(buf: bytes) -> dict:
    if len(buf) < 27 or not buf.startswith(b"Kaydara FBX Binary"):
        raise FbxGlobalSettingsError(
            "Файл не является бинарным FBX (нет сигнатуры 'Kaydara FBX Binary'). ASCII FBX не поддерживается.")
    c = Cursor(buf, 23)
    format_version = c.u32()
    if not (7400 <= format_version < 7700):
        raise FbxGlobalSettingsError(f"Версия формата FBX {format_version} не входит в поддерживаемый диапазон.")
    c.offset = 27
    null_rec = 25 if format_version >= 7500 else 13
    global_settings = None
    while c.offset < len(buf) - null_rec:
        node = _read_node(c, format_version, 0)
        if node is None:
            break
        if node["name"] == "GlobalSettings":
            global_settings = node
            break
        if node["name"] in ("Objects", "Connections"):
            break
    if global_settings is None:
        raise FbxGlobalSettingsError("В файле не найден узел GlobalSettings.")
    p70 = next((n for n in global_settings["children"] if n["name"] == "Properties70"), None)
    if p70 is None:
        raise FbxGlobalSettingsError("В GlobalSettings нет Properties70.")
    settings = {}
    for p in p70["children"]:
        if p["name"] != "P" or not p["props"]:
            continue
        settings[p["props"][0]] = p["props"][4:]

    def num(key) -> Optional[float]:
        arr = settings.get(key)
        return float(arr[0]) if arr and arr[0] is not None else None

    result = {
        "format_version": format_version,
        "unit_scale_factor": num("UnitScaleFactor"),
        "up_axis": num("UpAxis"),
        "up_axis_sign": num("UpAxisSign"),
        "front_axis": num("FrontAxis"),
        "front_axis_sign": num("FrontAxisSign"),
        "coord_axis": num("CoordAxis"),
        "coord_axis_sign": num("CoordAxisSign"),
    }
    if not result["unit_scale_factor"] or result["unit_scale_factor"] <= 0:
        raise FbxGlobalSettingsError("GlobalSettings.UnitScaleFactor отсутствует или некорректен.")
    for key in ("up_axis", "up_axis_sign", "front_axis", "front_axis_sign", "coord_axis", "coord_axis_sign"):
        if result[key] is None:
            raise FbxGlobalSettingsError(f"GlobalSettings.{key} отсутствует.")
    result["mm_per_unit"] = result["unit_scale_factor"] * 10
    return result


def assert_supported_axis_profile(settings: dict) -> None:
    p = SUPPORTED_AXIS_PROFILE
    if (int(settings["up_axis"]) != p["up_axis"] or int(settings["up_axis_sign"]) != p["up_axis_sign"]
            or int(settings["front_axis"]) != p["front_axis"] or int(settings["front_axis_sign"]) != p["front_axis_sign"]
            or int(settings["coord_axis"]) != p["coord_axis"] or int(settings["coord_axis_sign"]) != p["coord_axis_sign"]):
        raise FbxGlobalSettingsError(
            "Комбинация осей FBX не поддерживается в этой версии импорта — поддержан только профиль первого приёмочного файла.")
